### PR TITLE
GLAS height metrics

### PR DESCRIPTION
**Pull request by tcchiao**
Adding functions to calculate all height metrics listed in [Harris et al's spread sheet](https://docs.google.com/spreadsheets/d/1Hb67l3xYCfgxKu9TnpbEfY6iQo4ISNEvQewpmYvH9yQ/edit#gid=1090156694). Also adding functionalities to call necessary functions based on the param names.

### CODE REVIEW

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 1]
```diff
@@ -61,29 +280,328 @@ def get_ground_peak_distance(bins, wf, buffer=1):
...
+def highest_energy_value(ds):
+    """
+    Highest energy value in the waveform
+    """
+    return ds.processed_wf.max(dim="rec_bin")
+
+
+def wf_variance(ds):
+    """
+    Variance of the waveform
```
Comment (orianac) [inline carbonplan_trace/v1/glas_height_metrics.py:430]: why weighting it?

File: carbonplan_trace/v1/glas_preprocess.py [hunk 2]
```diff
@@ -102,39 +159,43 @@ def find_gaussian_fit_sigma(wf, default=3):
...
+def select_valid_area(bins, wf, signal_begin_dist, signal_end_dist):
     """
     vectorized
     """
     # get mask of valid area
-    valid = (bins > signal_begin_distance) & (bins < signal_end_distance)
+    valid = (bins > signal_begin_dist) & (bins < signal_end_dist)
 
     # set all invalid area to 0s and clip lower at 0
     wf = wf.where(valid, other=0).clip(min=0)
```
Comment (orianac) [inline carbonplan_trace/v1/glas_preprocess.py:184]: why is clipping necessary? aren't they set to zeros?

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 3]
```diff
@@ -61,29 +280,328 @@ def get_ground_peak_distance(bins, wf, buffer=1):
...
+        wf=ds.processed_wf,
+        signal_begin_dist=ds.sig_begin_dist,
+        signal_end_dist=ds.start_of_ground_peak_dist,
+    )
+
+    # make sure dimensions matches up
+    dims = ds.processed_wf.dims
+    sig_beg_to_ground = sig_beg_to_ground.transpose(dims[0], dims[1], dims[2])
+    # reverse denoise since the total energy is also including noise
+    sig_beg_to_ground = sig_beg_to_ground + ds.noise_mean
```
Comment (orianac) [inline carbonplan_trace/v1/glas_height_metrics.py:511]: is noise_mean calculated based upon the full waveform or is it specific to the sig_beg to ground - does it need to be recalculated for that slice? also just curious about whether the different noise levels could alter this calculation depending on the shot?

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 4]
```diff
@@ -61,29 +280,328 @@ def get_ground_peak_distance(bins, wf, buffer=1):
...
+
+    # total energy of the raw waveform
+    total = ds.rec_wf.sum(dim="rec_bin")
+
+    return sig_beg_to_ground.sum(dim="rec_bin") / total
+
+
+def pct_canopy_cover(ds, cutoff_height=1.0):
```
Comment (orianac) [inline carbonplan_trace/v1/glas_height_metrics.py:519]: i see in the spreadsheet that cutoff is arbitrary - maybe we could pass some broad mapping based on a landcover classification- feels like it could be sensitive in dense tropical or mixed forests.

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 5]
```diff
@@ -61,29 +280,328 @@ def get_ground_peak_distance(bins, wf, buffer=1):
...
+    vegetation_returns = select_valid_area(
+        bins=ds.rec_wf_sample_dist,
+        wf=ds.processed_wf,
+        signal_begin_dist=ds.sig_begin_dist,
+        signal_end_dist=ds.ground_peak_dist - cutoff_height,
+    )
+
+    # make sure dimensions matches up
+    dims = ds.processed_wf.dims
+    vegetation_returns = vegetation_returns.transpose(dims[0], dims[1], dims[2])
```
Comment (orianac) [inline carbonplan_trace/v1/glas_height_metrics.py:540]: for consistency, why doesn't it have to be re-noised like in L511 above?

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 6]
```diff
@@ -134,4 +652,94 @@ def plot_shot(record):
...
+    return ds
+
+
+def get_heights_from_distance(ds, top_metric, bottom_metric):
+    # check if the metric is in input ds, recalculate if not
+    for metric in [top_metric, bottom_metric]:
+        ds = get_dist_metric_value(ds, metric)
+
+    # multiply with -1 since distance is measured from satellite to object, thus top has a smaller value
+    return -1 * (ds[top_metric] - ds[bottom_metric])
```
Comment (orianac) [inline carbonplan_trace/v1/glas_height_metrics.py:698]: could we add some test in here for negative heights? because all of this should result in positive numbers, right?

File: carbonplan_trace/v1/glas_preprocess.py [hunk 7]
```diff
@@ -62,23 +72,70 @@ def process_coordinates(lat, lon, time):
...
     mask = ~ds.lat.isnull()
     mask.name = 'mask'
+    m1 = get_pct_false(mask)
+    print(f'filtering out {m1}% of records due to null GLAH14 data')
+
+    # Harris et al 2021 filtering conditions listed in Supplementary Information
+    mask = mask & (ds.num_gaussian_peaks >= 2)  # have at least two peaks
+    m2 = get_pct_false(mask)
+    print(f'filtering out {m2-m1}% of records due to number of gaussian peaks')
+    m1 = m2
```
Comment (orianac) [inline carbonplan_trace/v1/glas_preprocess.py:98]: this section feels complicated- perhaps just an extra line in the docstring about the stepwise filtering would add clarification

File: carbonplan_trace/v1/glas_preprocess.py [hunk 8]
```diff
@@ -62,23 +72,70 @@ def process_coordinates(lat, lon, time):
...
+        abs(ds.elevation_SRTM - (ds.elevation + ds.elevation_correction)) <= 30
+    )  # discrepancy bt SRTM and GLAS derived elevation less than 30m
+    m2 = get_pct_false(mask)
+    print(f'filtering out {m2-m1}% of records due to discrepancy in SRTM and GLAS elev')
+    m1 = m2
+
+    mask = (
+        mask
+        # signal beginning is less than 70m (otherwise indicates potential inference of signal)
+        & (abs(ds.ground_peak_dist - ds.sig_begin_dist) <= 70)
```
Comment (orianac) [inline carbonplan_trace/v1/glas_preprocess.py:117]: just to confirm, these subtractions are before they've been flipped into heights, right?

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 9]
```diff
@@ -1,41 +1,260 @@
...
+
+def pct_25_to_ground_ht(ds):
+    return get_heights_from_distance(ds, top_metric='pct_25_dist', bottom_metric='ground_peak_dist')
+
+
+def pct_50_to_ground_ht(ds):
+    return get_heights_from_distance(ds, top_metric='pct_50_dist', bottom_metric='ground_peak_dist')
+
+
+def pct_75_to_ground_ht(ds):
```
[the remark this answers is not in the document]
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:123]: Perhaps a TODO

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 10]
```diff
@@ -61,29 +280,328 @@ def get_ground_peak_distance(bins, wf, buffer=1):
...
+    return ds.ground_peak_dist - ground_peak_trailing_extent
 
-    return ds
+
+def get_adj_ground_peak_dist(ds):
+    """
+    the centroid position of whichever of the two lowest fitted Gaussian peaks has greater amplitude, as defined by Rosette, North, and Suarez (2008)
+    """
+    # find the larger peak between the bottom two
+    loc = ds.gaussian_amp.isel(n_gaussian_peaks=slice(2)).fillna(0).argmax(dim="n_gaussian_peaks")
```
[the remark this answers is not in the document]
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:301]: We have a filter where we only process records with at least 2 peaks -- fillna is needed here because argmax doesn't deal with all nans.

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 11]
```diff
@@ -61,29 +280,328 @@ def get_ground_peak_distance(bins, wf, buffer=1):
...
+    # the value of nsig is coded based on the GLAS Algorithm Theoretical Basis Document retrieved at
+    # https://www.csr.utexas.edu/glas/pdf/WFAtbd_v5_02011Sept.pdf  (See Appendix 3, pg 99)
+    time_of_switch = datetime(2000, 1, 1, 12, 0, 0, tzinfo=timezone.utc).timestamp() + 289742400
+    # for any time before the time of switch, b_nsig = 3.5, 7.5 afterwards
+    b_nsig = xr.where(ds.time < time_of_switch, x=3.5, y=7.5)
+
+    sig_begin_amp = ds.noise_mean + b_nsig * ds.noise_sd
+
+    # calculate slope as y2-y1 / x2-x1
+    fslope_WHRC = (canopy_amp - sig_begin_amp) / (canopy_dist - ds.sig_begin_dist)
```
[the remark this answers is not in the document]
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:417]: double check and filter off if not (add another filtering step after height metrics)

File: carbonplan_trace/v1/glas_height_metrics.py (hunk 3 above)
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:511]: switch to processed waveform

File: carbonplan_trace/v1/glas_height_metrics.py (hunk 6 above)
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:698]: yes

File: carbonplan_trace/v1/glas_preprocess.py [hunk 12]
```diff
@@ -62,23 +72,70 @@ def process_coordinates(lat, lon, time):
...
 
-    return x, y, time_out
+    return ds
 
 
 def get_mask(ds):
     """
     True in mask = records to use
     """
+
```
Comment (tcchiao) [inline carbonplan_trace/v1/glas_preprocess.py:84]: add citation to the harris et al paper

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 13]
```diff
@@ -1,41 +1,260 @@
...
+    """
+    Obtained directly from GLAH14 data
+    """
+    # calculate the bias between reference range to the bottom of received wf
+    ref_range_bias = ds.rec_wf_sample_dist.max(dim="rec_bin") - ds.ref_range
+    return ds.gaussian_mu + ds.ref_range + ref_range_bias
+
+
+def get_percentile_dist(ds, percentile):
     """
```
[the remark this answers is not in the document]
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:230]: added comment

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 14]
```diff
@@ -1,41 +1,260 @@
...
+    # lowest elevation = largest distance
+    return ds.gaussian_fit_dist.max(dim="n_gaussian_peaks")
 
 
-def get_ground_peak_distance(bins, wf, buffer=1):
+def get_ground_peak_dist_Sun(ds, buffer=1):
     """
-    bins and wf are expected to be data arrays with rec_bin, record_index, shot_number as dims
-    buffer is the minimum index where the ground peak can be located
+    Identify lowest peak in smoothed waveform, adopted after Sun et al 2008
```
[the remark this answers is not in the document]
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:251]: added comment

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 15]
```diff
@@ -61,29 +280,328 @@ def get_ground_peak_distance(bins, wf, buffer=1):
     return distance
```
[the remark this answers is not in the document]
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:280]: added

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 16]
```diff
@@ -61,29 +280,328 @@ def get_ground_peak_distance(bins, wf, buffer=1):
...
+    """
+    Mean height of the waveform from the signal beginning to signal end
+    """
+    wf = ds.processed_wf
+    bins = ds.rec_wf_sample_dist
+
+    return bins.weighted(wf).mean("rec_bin")
+
+
+def get_05th_pct_dist(ds):
```
[the remark this answers is not in the document]
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:327]: added todo

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 17]
```diff
@@ -61,29 +280,328 @@ def get_ground_peak_distance(bins, wf, buffer=1):
...
+    """
+    wf = ds.processed_wf
+    bins = ds.rec_wf_sample_dist
+
+    return bins.isel(rec_bin=wf.argmax(dim="rec_bin"))
+
+
+def get_leading_edge_dist(ds):
+    """
+    the first elevation at which the waveform is half of the maximum signal above the background noise value
```
[the remark this answers is not in the document]
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:379]: leading edge extent was used in filtering conditions, you can see the figure in https://daac.ornl.gov/NACP/guides/NACP_Boreal_Biome_Biomass.html for a good visualization

File: carbonplan_trace/v1/glas_height_metrics.py [hunk 18]
```diff
@@ -61,29 +280,328 @@ def get_ground_peak_distance(bins, wf, buffer=1):
...
+    canopy_amp = ds.gaussian_amp.isel(n_gaussian_peaks=canopy_dist)
+
+    # calculate amplitude at signal begin as noise mean + nsig * noise sd
+    # the value of nsig is coded based on the GLAS Algorithm Theoretical Basis Document retrieved at
+    # https://www.csr.utexas.edu/glas/pdf/WFAtbd_v5_02011Sept.pdf  (See Appendix 3, pg 99)
+    time_of_switch = datetime(2000, 1, 1, 12, 0, 0, tzinfo=timezone.utc).timestamp() + 289742400
+    # for any time before the time of switch, b_nsig = 3.5, 7.5 afterwards
+    b_nsig = xr.where(ds.time < time_of_switch, x=3.5, y=7.5)
+
+    sig_begin_amp = ds.noise_mean + b_nsig * ds.noise_sd
```
[the remark this answers is not in the document]
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:414]: the amplitude is not available in GLAH14 data, but added comments to clarify

File: carbonplan_trace/v1/glas_height_metrics.py (hunk 1 above)
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:430]: we're getting the variance on distance instead of waveform amplitude here.

File: carbonplan_trace/v1/glas_height_metrics.py (hunk 4 above)
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:519]: yup we could, holding until we asked Mary about her tree cover data source

File: carbonplan_trace/v1/glas_height_metrics.py (hunk 5 above)
Comment (tcchiao) [reply carbonplan_trace/v1/glas_height_metrics.py:540]: since we're already using the smoothed/denoised waveform here.

File: carbonplan_trace/v1/glas_preprocess.py (hunk 8 above)
Comment (tcchiao) [reply carbonplan_trace/v1/glas_preprocess.py:117]: yup they're on distance, but the `abs` should take care of the sign anyways

File: carbonplan_trace/v1/glas_preprocess.py (hunk 2 above)
Comment (tcchiao) [reply carbonplan_trace/v1/glas_preprocess.py:184]: the clipping is for within the valid area. Gaussian smoothing can't handle negative numbers so clipping to zero
